RenderStatePackager/src/RenderStatePackager.cpp

```diff
@@ -33,217 +33,314 @@ namespace Diligent
-        };
-
-        for (Uint32 PipelineID = 0; PipelineID < ParserInfo.GraphicsPipelineStateCount; PipelineID++)
-        {
-            auto pResourceDescRSN = pNotationParser->GetGraphicsPipelineStateByIndex(PipelineID);
-
-            GraphicsPipelineStateCreateInfo ResourceDesc = {};
-            UnpackPipelineStateCreateInfo(*pResourceDescRSN, ResourceDesc);
-            ResourceDesc.GraphicsPipeline             = static_cast<GraphicsPipelineDesc>(pResourceDescRSN->Desc);
-            ResourceDesc.GraphicsPipeline.pRenderPass = FindRenderPass(pResourceDescRSN->pRenderPassName);
-
-            ResourceDesc.pVS = FindShader(pResourceDescRSN->pVSName);
-            ResourceDesc.pPS = FindShader(pResourceDescRSN->pPSName);
-            ResourceDesc.pDS = FindShader(pResourceDescRSN->pDSName);
-            ResourceDesc.pHS = FindShader(pResourceDescRSN->pHSName);
-            ResourceDesc.pGS = FindShader(pResourceDescRSN->pGSName);
-            ResourceDesc.pAS = FindShader(pResourceDescRSN->pASName);
-            ResourceDesc.pMS = FindShader(pResourceDescRSN->pMSName);
-
-            PipelineStateArchiveInfo ArchiveInfo = {};
-            ArchiveInfo.DeviceFlags              = m_DeviceBits;
-
-            if (!pArchive->AddGraphicsPipelineState(ResourceDesc, ArchiveInfo))
-                LOG_ERROR_AND_THROW("Failed to archive graphics pipeline '", ResourceDesc.PSODesc.Name, "'.");
-        }
+        m_pThreadPool->WaitForAllTasks();
 
-        for (Uint32 PipelineID = 0; PipelineID < ParserInfo.ComputePipelineStateCount; PipelineID++)
-        {
-            auto                           pResourceDescRSN = pNotationParser->GetComputePipelineStateByIndex(PipelineID);
-            ComputePipelineStateCreateInfo ResourceDesc     = {};
-
-            UnpackPipelineStateCreateInfo(*pResourceDescRSN, ResourceDesc);
-            ResourceDesc.pCS = FindShader(pResourceDescRSN->pCSName);
+        for (auto& Exception : Exceptions)
+            if (Exception)
+                std::rethrow_exception(Exception);
+    }
+    catch (...)
+    {
+        return false;
+    }
+    return true;
+}
 
-            PipelineStateArchiveInfo ArchiveInfo = {};
-            ArchiveInfo.DeviceFlags              = m_DeviceBits;
+bool RenderStatePackager::Execute(RefCntAutoPtr<IArchiver> pArchive)
+{
+    DEV_CHECK_ERR(pArchive != nullptr, "pArchive must not be null");
 
-            if (!pArchive->AddComputePipelineState(ResourceDesc, ArchiveInfo))
-                LOG_ERROR_AND_THROW("Failed to archive compute pipeline '", ResourceDesc.PSODesc.Name, "'.");
-        }
+    if (pArchive == nullptr)
+        return false;
 
-        for (Uint32 PipelineID = 0; PipelineID < ParserInfo.TilePipelineStateCount; PipelineID++)
+    try
+    {
+        std::vector<std::exception_ptr>        Exceptions;
+        std::vector<RefCntAutoPtr<IAsyncTask>> Tasks;
+        for (auto const& pNotationParser : m_RSNParsers)
```

Comment:
I suggest you rework this without using mutices. You create
`vector<vector<RefCntAutoPtr<IShader>>> Shaders(m_RSNParsers.size());`
and then add shaders to the respective vector for each task without the mutex.
After all tasks are finished, you merge all shaders into single array m_Shaders